Fixed #30020 -- Fixed reading nulls with LayerMapping.

flibbertigibbet opened the pull request:
## Overview

Ticket [30020](https://code.djangoproject.com/ticket/30020).

Added a GDAL function wrapper to recognize if a field is either null or "unset".

Modified `LayerMapping` importer to import null or unset GDAL values as null. Previously, null numeric values in an input file would be set as zeroes on the destination model, and null strings would be set as empty strings.

## Testing notes

I've run the `gis_tests` suite successfully against PostGIS, Spatialite, and MySQL database backends. For each, tests succeeded with both python 2.7 / GDAL 2.2, and with python 3.6 / GDAL 2.1.

I made a small demo app [here](https://github.com/flibbertigibbet/django-layermapping-null-demo/). The demo app is in Django 1.11, and can run its tests against a separate 1.11 backport branch with the commits here cherry-picked.

The demo app repository can also run the core Django tests with Docker containers; there are branches for the different database backends. I used it to run the core GIS suite of tests against the Django master-based branch here.

claudep commented:
Do you think it would be possible to replace the Shapefile test data with a GeoJSON one? Or is the null issue really specific to shapefiles? Having a text instead of binary data source would be nice for readability.

claudep commented:
Wow, great! Existing tests are based on Shapefile probably for historical reasons, I guess GeoJSON layers were not supported in the past.
Now do you think it's worth keeping the Shapefile tests in this patch? In other words, is anything in the patch specific to Shapefiles?

claudep commented:
By the way, feel free to add your name to the AUTHORS file.

flibbertigibbet commented:
I removed the test Shapefile. Switching the `gdal_tests/test_ds.py` data source added here to use the GeoJSON driver instead of the Shapefile driver surfaced an unrelated issue with reading numbers from GeoJSON, in which `OFTReal` typed fields were treated as `OFTInteger` because they had no precision set.

I removed the logic to cast reals without precision to integers in 2b9faa0, which fixed the issue and did not break any existing tests. It's unclear why that logic was in place, as there is no comment to explain the choice, and the commit history that adds it predates the repository move from SVN to git.

claudep commented on django/contrib/gis/gdal/field.py:
```diff
@@ -34,34 +34,37 @@ def __init__(self, feat, index):
     def as_double(self):
         "Retrieve the Field's value as a double (float)."
+        if not capi.is_field_set(self._feat.ptr, self._index):
+            return None
```
It looks like this pattern would benefit to be applied as a method decorator, as that would limit code duplication.

claudep commented on tests/gis_tests/data/has_nulls/has_nulls.geojson:
```diff
@@ -0,0 +1,10 @@
+{ "type": "Feature", "properties": { "UUID": "fa2ba67c-a135-4338-b924-a9622b5d869f", "FILL_COLOR": "#dfc827", "FILL_OPACI": 0.5, "SOME_NUMBE": null }, "geometry": { "type": "Polygon", "coordinates": [ [ [ -75.112152099609375, 39.94870062390347 ], [ -75.02838134765625, 39.990799335838034 ], [ -74.959716796875, 39.920269337633975 ], [ -75.025634765625, 39.846503602417698 ], [ -75.12725830078125, 39 ...
+{ "type": "Feature", "properties": { "UUID": "4494c1f3-55ab-4256-b365-12115cb388d5", "FILL_COLOR": "#982424", "NAME": "north", "FILL_OPACI": 0.8, "SOME_NUMBE": 0.0 }, "geometry": { "type": "Polygon", "coordinates": [ [ [ -75.249481201171875, 39.993955693973312 ], [ -75.274200439453125, 40.050745194796299 ], [ -75.205535888671875, 40.119040222688795 ], [ -75.081939697265625, 40.032872116772268 ],  ...
+]
+}
```
Wonder what's the opinion of our fellows, but it might be nicer to have an indented output:
```
{
  "name": "has_nulls",
  "type": "FeatureCollection",
  "crs": {
    "properties": {
      "name": "urn:ogc:def:crs:OGC:1.3:CRS84"
    },
    "type": "name"
  },
  "features": [
    {
      "geometry": {
        "coordinates": [
          [
            [
              -75.14923095703125,
              39.8928799002948
            ],
            [
              -75.25772094726562,
              39.889718875996685
            ],
            [
              -75.201416015625,
              40.03076909369051
            ],
            [
              -75.07644653320312,
              39.99500778093748
            ],
            [
              -75.14785766601562,
              39.95291166179976
            ],
            [
              -75.135498046875,
              39.91605629078665
            ],
            [
              -75.14923095703125,
              39.8928799002948
            ]
          ]
        ],
        "type": "Polygon"
      },
      "properties": {
        "UUID": "1378c26f-cbe6-44b0-929f-eb330d4991f5",
        "FILL_COLOR": "#4639bc",
        "FILL_OPACI": 0.5,
        "SOME_NUMBE": 1.001,
        "NAME": "Philadelphia"
      },
      "type": "Feature"
    },
    {
      "geometry": {
        "coordinates": [
          [
            [
              -75.11215209960938,
              39.94870062390347
            ],
            [
              -75.02838134765625,
              39.990799335838034
            ],
            [
              -74.959716796875,
              39.920269337633975
            ],
            [
              -75.025634765625,
              39.8465036024177
            ],
            [
              -75.12725830078125,
              39.871803651624425
            ],
            [
              -75.11215209960938,
              39.94870062390347
            ]
          ]
        ],
        "type": "Polygon"
      },
      "properties": {
        "UUID": "fa2ba67c-a135-4338-b924-a9622b5d869f",
        "FILL_COLOR": "#dfc827",
        "FILL_OPACI": 0.5,
        "SOME_NUMBE": null
      },
      "type": "Feature"
    },
    {
      "geometry": {
        "coordinates": [
          [
            [
              -75.24948120117188,
              39.99395569397331
            ],
            [
              -75.27420043945312,
              40.0507451947963
            ],
            [
              -75.20553588867188,
              40.119040222688795
            ],
            [
              -75.08193969726562,
              40.03287211677227
            ],
            [
              -75.16433715820312,
              39.97396296240704
            ],
            [
              -75.24948120117188,
              39.99395569397331
            ]
          ]
        ],
        "type": "Polygon"
      },
      "properties": {
        "UUID": "4494c1f3-55ab-4256-b365-12115cb388d5",
        "FILL_COLOR": "#982424",
        "FILL_OPACI": 0.8,
        "SOME_NUMBE": 0.0,
        "NAME": "north"
      },
      "type": "Feature"
    }
  ]
}
```

claudep commented on tests/gis_tests/gdal_tests/test_ds.py:
```diff
@@ -53,7 +53,27 @@
+            'NAME': OFTString,
+            'SOME_NUMBE': OFTReal,
+            'FILL_COLOR': OFTString,
+            'FILL_OPACI': OFTReal,
```
What about having an integer and a datetime content in the GeoJSON (instead of 2 strings and two reals)?

claudep commented on tests/gis_tests/layermap/models.py:
```diff
@@ -92,3 +115,11 @@ class Invalid(models.Model):
+                     'name': 'NAME',
+                     'fill_opacity': 'FILL_OPACI',
+                     'some_number': 'SOME_NUMBE',
+                     }
```
We tend to favor the dict formatting as:
```
some_dict = {
    'key1': 'value',
}
```

claudep commented:
Good progress, thanks!
@timgraham or @carltongibson, it may be time for another review. Would you cut the changes in more than one commit?

flibbertigibbet commented:
Sorry, but I'm not sure what "cut the changes in more than one commit" means. Would you like me to rebase down to one commit?

> Good progress, thanks!
> @timgraham or @carltongibson, it may be time for another review. Would you cut the changes in more than one commit?

claudep commented:
Rebasing to one commit isn't the issue, it can easily be done before the merge. However, one might argue that some changes in your patch should be committed separately. That's generally asked by the Django fellows at the final review stage, let's wait for that to happen.

claudep commented:
Kathryn, may I ask you to rebase your branch (after 0004daa53689)?
I think you may also squash to one commit. If you are asked later to separate commits, it will not match the current series anyway.

flibbertigibbet commented:
Rebased and squashed

timgraham commented:
I pushed some cosmetic edits and see that `test_null_number_imported_not_allowed` (where I changed `assertNotEqual(..., 3)` to `assertEqual()`) behaves differently on MySQL and Oracle (returning 1 instead of 0). Is that expected?

flibbertigibbet commented:
No, the different database backend behavior on failed save was unexpected. I ran into that as well, which is why I had the assert in `test_null_number_imported_not_allowed` only checking that not all records made it in.

timgraham commented:
Would you like to investigate that further to determine if the cause is a bug?

flibbertigibbet commented:
I don't expect it's related to this PR, so no, I wasn't intending to look into it any further.

timgraham commented:
It looks like the issue is that one failure e.g. `(1048, "Column 'some_datetime' cannot be null")` causes a cascading error the causes the rest of the objects to fail to load with: `An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.`

For some reason, the first object that loads successfully is only visible in the `count()` query in the test on MySQL and Oracle.

@claudep, do you think the cascading failure issue merits investigation in a separate ticket?

flibbertigibbet commented:
From the [discussion on ticket 11156](https://code.djangoproject.com/ticket/11156#comment:14), I'm guessing the save behavior is a result of differing atomic transaction management on the database backends.

timgraham commented:
Thanks for the research. I proposed `self.assertLessEqual(DoesNotAllowNulls.objects.count(), 1)` and added a comment about the issue.

timgraham commented:
Some failures on Ubuntu 16.04 (possibly due to GDAL 1.11 which we may drop support for in Django 3.0 as GDAL 1.11 was released over 5 years ago, but later versions of GDAL may also be affected):
```
======================================================================
ERROR: test_null_number_imported (gis_tests.layermap.tests.LayerMapTest)
LayerMapping import of GeoJSON with a null numeric value.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/media/sf_django/tests/gis_tests/layermap/tests.py", line 329, in test_null_number_imported
    lm = LayerMapping(HasNulls, has_nulls_geojson, has_nulls_mapping)
  File "/media/sf_django/django/contrib/gis/utils/layermapping.py", line 108, in __init__
    self.check_layer()
  File "/media/sf_django/django/contrib/gis/utils/layermapping.py", line 258, in check_layer
    (ogr_field, ogr_field.__name__, fld_name))
django.contrib.gis.utils.layermapping.LayerMapError: OGR field "<class 'django.contrib.gis.gdal.field.OFTString'>" (of type OFTString) cannot be mapped to Django DateTimeField.

======================================================================
ERROR: test_null_number_imported_not_allowed (gis_tests.layermap.tests.LayerMapTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/media/sf_django/tests/gis_tests/layermap/tests.py", line 371, in test_null_number_imported_not_allowed
    lm = LayerMapping(DoesNotAllowNulls, has_nulls_geojson, has_nulls_mapping)
  File "/media/sf_django/django/contrib/gis/utils/layermapping.py", line 108, in __init__
    self.check_layer()
  File "/media/sf_django/django/contrib/gis/utils/layermapping.py", line 258, in check_layer
    (ogr_field, ogr_field.__name__, fld_name))
django.contrib.gis.utils.layermapping.LayerMapError: OGR field "<class 'django.contrib.gis.gdal.field.OFTString'>" (of type OFTString) cannot be mapped to Django DateTimeField.

======================================================================
ERROR: test_null_string_imported (gis_tests.layermap.tests.LayerMapTest)
Test LayerMapping import of GeoJSON with a null string value.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/media/sf_django/tests/gis_tests/layermap/tests.py", line 337, in test_null_string_imported
    lm = LayerMapping(HasNulls, has_nulls_geojson, has_nulls_mapping)
  File "/media/sf_django/django/contrib/gis/utils/layermapping.py", line 108, in __init__
    self.check_layer()
  File "/media/sf_django/django/contrib/gis/utils/layermapping.py", line 258, in check_layer
    (ogr_field, ogr_field.__name__, fld_name))
django.contrib.gis.utils.layermapping.LayerMapError: OGR field "<class 'django.contrib.gis.gdal.field.OFTString'>" (of type OFTString) cannot be mapped to Django DateTimeField.

======================================================================
ERROR: test_nullable_boolean_imported (gis_tests.layermap.tests.LayerMapTest)
LayerMapping import of GeoJSON with a nullable boolean value.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/media/sf_django/tests/gis_tests/layermap/tests.py", line 346, in test_nullable_boolean_imported
    lm = LayerMapping(HasNulls, has_nulls_geojson, has_nulls_mapping)
  File "/media/sf_django/django/contrib/gis/utils/layermapping.py", line 108, in __init__
    self.check_layer()
  File "/media/sf_django/django/contrib/gis/utils/layermapping.py", line 258, in check_layer
    (ogr_field, ogr_field.__name__, fld_name))
django.contrib.gis.utils.layermapping.LayerMapError: OGR field "<class 'django.contrib.gis.gdal.field.OFTString'>" (of type OFTString) cannot be mapped to Django DateTimeField.

======================================================================
ERROR: test_nullable_datetime_imported (gis_tests.layermap.tests.LayerMapTest)
LayerMapping import of GeoJSON with a nullable date/time value.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/media/sf_django/tests/gis_tests/layermap/tests.py", line 354, in test_nullable_datetime_imported
    lm = LayerMapping(HasNulls, has_nulls_geojson, has_nulls_mapping)
  File "/media/sf_django/django/contrib/gis/utils/layermapping.py", line 108, in __init__
    self.check_layer()
  File "/media/sf_django/django/contrib/gis/utils/layermapping.py", line 258, in check_layer
    (ogr_field, ogr_field.__name__, fld_name))
django.contrib.gis.utils.layermapping.LayerMapError: OGR field "<class 'django.contrib.gis.gdal.field.OFTString'>" (of type OFTString) cannot be mapped to Django DateTimeField.

======================================================================
ERROR: test_uuids_imported (gis_tests.layermap.tests.LayerMapTest)
LayerMapping import of GeoJSON with UUIDs.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/media/sf_django/tests/gis_tests/layermap/tests.py", line 362, in test_uuids_imported
    lm = LayerMapping(HasNulls, has_nulls_geojson, has_nulls_mapping)
  File "/media/sf_django/django/contrib/gis/utils/layermapping.py", line 108, in __init__
    self.check_layer()
  File "/media/sf_django/django/contrib/gis/utils/layermapping.py", line 258, in check_layer
    (ogr_field, ogr_field.__name__, fld_name))
django.contrib.gis.utils.layermapping.LayerMapError: OGR field "<class 'django.contrib.gis.gdal.field.OFTString'>" (of type OFTString) cannot be mapped to Django DateTimeField.

======================================================================
FAIL: test03a_layers (gis_tests.gdal_tests.test_ds.DataSourceTest)
Testing Data Source Layers.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/media/sf_django/tests/gis_tests/gdal_tests/test_ds.py", line 164, in test03a_layers
    self.assertEqual(fld_value, layer.get_fields(fld_name))
AssertionError: Lists differ: [5, None, 8] != ['5', None, '8']

First differing element 0:
5
'5'

- [5, None, 8]
+ ['5', None, '8']
?  + +        + +


======================================================================
FAIL: test04_features (gis_tests.gdal_tests.test_ds.DataSourceTest)
Testing Data Source Features.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/media/sf_django/tests/gis_tests/gdal_tests/test_ds.py", line 237, in test04_features
    self.assertIsInstance(feat[k], v)
AssertionError: <django.contrib.gis.gdal.field.OFTReal object at 0x7f459af445f8> is not an instance of <class 'django.contrib.gis.gdal.field.OFTInteger'>
```

flibbertigibbet commented:
It looks like those were failures with the existing date/time field mapping, which was not added here, but I did add tests for it here; there were no tests for date/time field mappings in GDAL/LayerMapping previously, as far as I could tell. I'l try to reproduce this locally.

timgraham commented:
I confirmed that the failures aren't present with GDAL 2.0 so I think we can simply drop support for GDAL 1.11 (https://github.com/django/django/pull/10959).

flibbertigibbet commented:
I was able to reproduce the test failures with GDAL 1.11. The `DateTime` failures are due to the [GDAL GeoJSON driver](https://www.gdal.org/drv_geojson.html) not adding support for `DateTime` until version 2.0, and so can be addressed with GDAL version checks on the tests.

I have a [branch here](https://github.com/flibbertigibbet/django/tree/bugfix/fix-gis-tests-gdal-11) that addresses most of the test failures, but not all. There seems to be an issue with GDAL 1.11 integer/real type identification in GeoJSON I haven't been able to pin down.
